presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java

```diff
@@ -342,4 +351,187 @@ public void close()
+        private final Map<LifespanAndStage, LifespanAndStageState> committedRecoverableLifespanAndStages = new HashMap<>();
+
+        private final LifespanCommitter lifespanCommitter;
+        private final JsonCodec<TableCommitContext> tableCommitContextCodec;
+
+        LifespanAndStageStateTracker(LifespanCommitter lifespanCommitter, JsonCodec<TableCommitContext> tableCommitContextCodec)
+        {
+            this.lifespanCommitter = requireNonNull(lifespanCommitter, "lifespanCommitter is null");
+            this.tableCommitContextCodec = requireNonNull(tableCommitContextCodec, "tableCommitContextCodec is null");
+        }
+
+        public void update(Page page)
+        {
+            TableCommitContext tableCommitContext = getTableCommitContext(page);
+            LifespanAndStage lifespanAndStage = LifespanAndStage.fromTableCommitContext(tableCommitContext);
+            if (committedRecoverableLifespanAndStages.containsKey(lifespanAndStage)) {
+                return;
+            }
+
+            if (!tableCommitContext.isLifespanCommitRequired()) {
+                unrecoverableLifespanAndStageStates.putIfAbsent(lifespanAndStage, new LifespanAndStageState());
+                unrecoverableLifespanAndStageStates.get(lifespanAndStage).update(page, false);
+                return;
+            }
+
+            uncommittedRecoverableLifespanAndStageStates.putIfAbsent(lifespanAndStage, new HashMap<>());
+            Map<Integer, LifespanAndStageState> lifespanStageStatesPerTask = uncommittedRecoverableLifespanAndStageStates.get(lifespanAndStage);
+            lifespanStageStatesPerTask.putIfAbsent(tableCommitContext.getTaskId(), new LifespanAndStageState());
+            lifespanStageStatesPerTask.get(tableCommitContext.getTaskId()).update(page, true);
+
+            if (tableCommitContext.isLastPage()) {
+                checkState(!committedRecoverableLifespanAndStages.containsKey(lifespanAndStage), "LifespanAndStage already finished");
+                LifespanAndStageState lifespanAndStageState = lifespanStageStatesPerTask.get(tableCommitContext.getTaskId());
+                committedRecoverableLifespanAndStages.put(lifespanAndStage, lifespanAndStageState);
+                uncommittedRecoverableLifespanAndStageStates.remove(lifespanAndStage);
+                lifespanCommitter.commitLifespan(lifespanAndStage.getLifespan().getId(), lifespanAndStageState.getFragments());
+            }
+        }
+
+        List<Page> getStatisticsPagesToProcess(Page page)
+        {
+            TableCommitContext tableCommitContext = getTableCommitContext(page);
+            LifespanAndStage lifespanAndStage = LifespanAndStage.fromTableCommitContext(tableCommitContext);
+            if (!tableCommitContext.isLifespanCommitRequired()) {
+                return extractStatisticsRows(page).map(ImmutableList::of).orElse(ImmutableList.of());
+            }
+            if (!committedRecoverableLifespanAndStages.containsKey(lifespanAndStage)) {
```

Comment:
Do you mean `unrecoverable`? I think `uncommittedRecoverableLifespanAndStageStates` should contain the `lifespanAndStage` because it is just added to the map in the `update` phase.

Reply:
Yes, that's what i meant. But feel free to leave it as is if you don't think it makes much sense.